torch/csrc/distributed/c10d/init.cpp

```diff
@@ -38,6 +38,49 @@
 
 #include <torch/custom_class.h>
 
+namespace {
+
+// Wrapper to ensure GIL is released before destructing ProcessGroupGloo
+// TODO: move this somewhere more generally useful
+template <typename T>
+class IntrusivePtrNoGilDestructor {
+  c10::intrusive_ptr<T> impl_;
+public:
+  IntrusivePtrNoGilDestructor() = default;
+  IntrusivePtrNoGilDestructor(const IntrusivePtrNoGilDestructor&) = default;
+  IntrusivePtrNoGilDestructor(IntrusivePtrNoGilDestructor&&) = default;
+  IntrusivePtrNoGilDestructor& operator=(const IntrusivePtrNoGilDestructor&) = default;
+  IntrusivePtrNoGilDestructor& operator=(IntrusivePtrNoGilDestructor&&) = default;
+  /* implicit */ IntrusivePtrNoGilDestructor(c10::intrusive_ptr<T> impl) : impl_(std::move(impl)) {}
+  // This ctor is very important; see
+  // https://github.com/pybind/pybind11/issues/2957
+  explicit IntrusivePtrNoGilDestructor(T* impl) : impl_(c10::intrusive_ptr<T>::reclaim(impl)) {
+    // Must incref because impl comes in with refcount zero
+    c10::raw::intrusive_ptr::incref(impl);
+  }
+  ~IntrusivePtrNoGilDestructor() {
+    if (impl_) {
```

Comment:
is it because we need to keep intrusive_ptr python-free (couldn't put this logic directly in intrusive_ptr), that we have to define a separate class specifically for the destructor on GIL?

Reply:
not just that, but if the destructor doesn't actually block and is quick to run, releasing the GIL does more harm than good (it costs performance)